server/src/main/java/org/opensearch/gateway/AsyncShardBatchFetch.java

```diff
@@ -0,0 +1,341 @@
+        ) {
+            super(Loggers.getLogger(logger, "_" + logKey), type);
+            this.batchSize = shardAttributesMap.size();
+            this.getException = getResponseException;
+            this.isEmpty = isEmptyResponse;
+            cache = new HashMap<>();
+            shardIdToArray = new HashMap<>();
+            arrayToShardId = new HashMap<>();
+            fillShardIdKeys(shardAttributesMap.keySet());
+            this.shardResponseClass = clazz;
+            this.responseConstructor = responseGetter;
+            this.shardsBatchDataGetter = shardsBatchDataGetter;
+            this.emptyResponseBuilder = emptyResponseBuilder;
+            this.failedShardHandler = failedShardHandler;
+            this.logger = logger;
+        }
+
+        @Override
+        public Map<String, ? extends BaseNodeEntry> getCache() {
+            return cache;
+        }
+
+        @Override
+        public void deleteShard(ShardId shardId) {
+            if (shardIdToArray.containsKey(shardId)) {
+                Integer shardIdIndex = shardIdToArray.remove(shardId);
+                for (String nodeId : cache.keySet()) {
+                    cache.get(nodeId).clearShard(shardIdIndex);
+                }
+            }
+        }
+
+        @Override
+        public Map<DiscoveryNode, T> getCacheData(DiscoveryNodes nodes, Set<String> failedNodes) {
+            fillReverseIdMap();
+            return super.getCacheData(nodes, failedNodes);
+        }
+
+        /**
+         * Build a reverse map to get shardId from the array index, this will be used to construct the response which
+         * PrimaryShardBatchAllocator or ReplicaShardBatchAllocator are looking for.
+         */
+        private void fillReverseIdMap() {
+            arrayToShardId.clear();
+            for (Map.Entry<ShardId, Integer> indexMapping : shardIdToArray.entrySet()) {
+                arrayToShardId.putIfAbsent(indexMapping.getValue(), indexMapping.getKey());
+            }
+        }
+
+        @Override
+        public void initData(DiscoveryNode node) {
+            cache.put(node.getId(), new NodeEntry<>(node.getId(), shardResponseClass, batchSize, getException, isEmpty));
+        }
+
+        /**
+         * Put the response received from data nodes into the cache.
+         * Get shard level data from batch, then filter out if any shards received failures.
+         * After that complete storing the data at node level and mark fetching as done.
+         *
+         * @param node     node from which we got the response.
+         * @param response shard metadata coming from node.
+         */
+        @Override
+        public void putData(DiscoveryNode node, T response) {
+            NodeEntry<V> nodeEntry = cache.get(node.getId());
+            Map<ShardId, V> batchResponse = shardsBatchDataGetter.apply(response);
+            filterFailedShards(batchResponse);
+            nodeEntry.doneFetching(batchResponse, shardIdToArray);
```

Comment:
why is batchResponse iterated 2 times, failed shard handling and done fetching could have been combined?

Reply:
Filtering of failedShards needs to be done in ShardBatchCache class. Failed shards are at a cache level.

Saving shard responses work needs to be done in NodeEntry class - at a node level. 

Though failed shards are not iterated again in NodeEntry class as they are removed in filterFailedShards method.